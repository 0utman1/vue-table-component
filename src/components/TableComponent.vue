<template>
    <div class="table-component">

        <div v-if="showFilter" class="table-component__filter">
            <input 
                class="table-component__filter__field" 
                type="text" 
                v-model="filter" 
                :placeholder="filterPlaceholder"
           >
            <a 
                v-if="filter"
                @click="filter = ''"
                class="table-component__filter__clear"
            >Ã—</a>
        </div>

        <div class="table-component__table-wrapper">
            <table :class="fullTableClass">
                <caption class="table-component__table__caption" role="alert" aria-live="polite">
                    {{ ariaCaption }}
                </caption>
                <thead>
                <tr>
                    <table-column-header
                        @click="changeSorting"
                        v-for="column in columns"
                        :key="column.properties.show"
                        :sort="sort"
                        :column="column"
                    />
                </tr>
                </thead>
                <tbody>
                <table-row
                    v-for="row in displayedRows"
                    :key="row.vueTableComponentInternalRowId"
                    :row="row"
                    :columns="columns"
                /></table-row>
                </tbody>
            </table>
        </div>

        <div v-if="displayedRows.length === 0" class="table-component__message">
            {{ filterNoResults }}
        </div>

        <div style="display:none;">
            <slot></slot>
        </div>
    </div>
</template>

<script>
    import Column from '../classes/Column';
    import expiringStorage from '../expiring-storage';
    import Row from '../classes/Row';
    import TableColumnHeader from './TableColumnHeader';
    import TableRow from './TableRow';
    import { pick } from 'lodash';
    import settings from '../settings';
    import { isArray, merge } from 'lodash';

    export default {
        components: {
            TableColumnHeader,
            TableRow,
        },

        props: {
            data: { required: true, type: Array },

            showFilter: { default: true },
            sortBy: { default: '', type: String },
            sortOrder: { default: 'desc', type: String },

            cacheId: { default: '' },
            cacheLifetime: { default: 5 },

            tableClass: { default: settings.tableClass },
            filterPlaceholder: { default: settings.filterPlaceholder },
            filterNoResults: { default: settings.filterNoResutls },
        },

        data: () => ({
            columns: [],
            rows: [],
            filter: '',
            sort: {
                fieldName: '',
                order: '',
            },
            localSettings: {},
        }),

        computed: {
            fullTableClass() {
                const extraClasses = isArray(this.tableClass) ? 
                    this.tableClass : 
                    [this.tableClass];

                return ['table-component__table', ...extraClasses];
            },

            ariaCaption() {
                if (this.sort.fieldName === '') {
                    return 'Table not sorted';
                }

                return `Table sorted by ${this.sort.fieldName} ` +
                    (this.sort.order === 'asc' ? '(ascending)' : '(descending)');
            },

            displayedRows() {
                return this.sortedRows.filter(row => row.passesFilter(this.filter));
            },

            sortedRows() {
                if (this.sort.fieldName === '') {
                    return this.rows;
                }

                if (this.columns.length === 0) {
                    return this.rows;
                }

                const sortColumn = this.getColumn(this.sort.fieldName);

                if (! sortColumn) {
                    return this.rows;
                }

                return this.rows.sort(sortColumn.getSortPredicate(this.sort.order, this.columns));
            },

            storageKey() {
                return `vue-table-component.${window.location.host}${window.location.pathname}${this.cacheId}`;
            },
        },

        watch: {
            filter() {
                this.saveState();
            },
        },

        mounted() {
            this.columns = this.$slots.default
                .filter(column => column.componentInstance)
                .map(column => pick(column.componentInstance, [
                    'show', 'label', 'dataType', 'sortable', 'sortOn', 'filterable', 'filterOn',
                ]))
                .map(columnProperties => new Column(columnProperties));

            let rowId = 0;

            this.rows = this.data
                .map(rowData => {
                    rowData.vueTableComponentInternalRowId = rowId++;
                    return rowData;
                })
                .map(rowData => new Row(rowData, this.columns));
        },

        created() {
            this.sort.fieldName = this.sortBy;
            this.sort.order = this.sortOrder;

            this.restoreState();
        },

        methods: {
            changeSorting(column) {
                if (this.sort.fieldName !== column.properties.show) {
                    this.sort.fieldName = column.properties.show;
                    this.sort.order = 'asc';
                }

                if (this.sort.fieldName === column.properties.show) {
                    this.sort.order = (this.sort.order === 'desc' ? 'asc' : 'desc');
                }

                this.saveState();
            },

            getColumn(columnName) {
                return this.columns.find(column => column.properties.show === columnName);
            },

            saveState() {
                expiringStorage.set(this.storageKey, pick(this.$data, ['filter', 'sort']), this.cacheLifetime);
            },

            restoreState() {
                const previousState = expiringStorage.get(this.storageKey);

                if (previousState === null) {
                    return;
                }

                this.filter = previousState.filter;
                this.sort = previousState.sort;

                this.saveState();
            },
        },
    };
</script>
