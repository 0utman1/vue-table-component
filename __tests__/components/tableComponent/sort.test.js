import TableComponent from '../../../src';
import Vue from 'vue/dist/vue.js';
import simulant from 'simulant';
import LocalStorageMock from '../../helpers/LocalStorageMock';

const localStorage = new LocalStorageMock();

window.localStorage = localStorage;

describe('Sortable tableComponent', () => {
    Vue.use(TableComponent);

    beforeEach(() => {
        localStorage.clear();
    });

    it('can sort the data with by a specific column', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort the data with by another specific column', async () => {
        setDocumentInnerHtml({ sortBy: 'lastName' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can sort the data with by a specific column in a specific order', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName', order: 'desc' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('it will change the sort order when clicking the header of the column with the active sort', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        const firstColumnHeader = document.getElementsByTagName('th')[0];

        await simulant.fire(firstColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will sort the data descendingly if the header of of column without the active sort is clicked', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        const secondColumnHeader = document.getElementsByTagName('th')[1];

        await simulant.fire(secondColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });
    
    it('will not sort on a column that is not sortable', async () => {
        setDocumentInnerHtml({ sortBy: 'actions', order: 'desc' });

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not sort data when clicking a non-sortable column header', async () => {
        setDocumentInnerHtml({ sortBy: 'firstName' });

        await createVm();

        const thirdColumnHeader = document.getElementsByTagName('th')[2];

        await simulant.fire(thirdColumnHeader, 'click');

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});

function setDocumentInnerHtml({ sortBy = '', order = '' } = {}) {

    document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        :data="[{ id: 1, firstName: 'John', lastName: 'Lennon', actions: 'Red' },
                                { id: 2, firstName: 'Paul', lastName: 'McCartney', actions: 'Blue' },
                                { id: 3, firstName: 'George', lastName: 'Harrison', actions: 'Green' },
                                { id: 4, firstName: 'Ringo', lastName: 'Starr', actions: 'Yellow' }]"
                        sort-by="${sortBy}"
                        sort-order="${order}"
                    >
                        <table-column show="firstName" label="First name"></table-column>
                        <table-column show="lastName" label="Last name"></table-column>
                        <table-column show="actions" label="Color" :sortable="false"></table-column>
                    </table-component>
                </div>
            </div>
        `;
}

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});
}
