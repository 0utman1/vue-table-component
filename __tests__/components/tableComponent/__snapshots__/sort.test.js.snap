// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sortable tableComponent can sort the data with by a specific column 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort-\\">
    First name
</th><th class=\\"table-component__th--sort\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr><tr><td>John</td><td>Lennon</td><td>Red</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableComponent can sort the data with by a specific column in a specific order 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort-desc\\">
    First name
</th><th class=\\"table-component__th--sort\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>John</td><td>Lennon</td><td>Red</td></tr><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableComponent can sort the data with by another specific column 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort-\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>John</td><td>Lennon</td><td>Red</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableComponent it will change the sort order when clicking the header of the column with the active sort 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort-desc\\">
    First name
</th><th class=\\"table-component__th--sort\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>John</td><td>Lennon</td><td>Red</td></tr><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableComponent will not sort data when clicking a non-sortable column header 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort-\\">
    First name
</th><th class=\\"table-component__th--sort\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr><tr><td>John</td><td>Lennon</td><td>Red</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableComponent will not sort on a column that is not sortable 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr><tr><td>John</td><td>Lennon</td><td>Red</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;

exports[`Sortable tableComponent will sort the data descendingly if the header of of column without the active sort is clicked 1`] = `
"
            <div id=\\"app\\"><div><div class=\\"table-component\\"><div class=\\"table-component__filter\\"><input type=\\"text\\" name=\\"table-component-filter\\" placeholder=\\"Filter table…\\" class=\\"table-component__filter__field\\"> <!----></div> <div class=\\"table-component__table-wrapper\\"><table class=\\"table-component__table\\"><thead><tr><th class=\\"table-component__th--sort\\">
    First name
</th><th class=\\"table-component__th--sort-desc\\">
    Last name
</th><th>
    Color
</th></tr></thead> <tbody><tr><td>John</td><td>Lennon</td><td>Red</td></tr><tr><td>George</td><td>Harrison</td><td>Green</td></tr><tr><td>Ringo</td><td>Starr</td><td>Yellow</td></tr><tr><td>Paul</td><td>McCartney</td><td>Blue</td></tr></tbody></table></div> <!----> <div style=\\"display: none;\\"><th>
    First name
</th> <th>
    Last name
</th> <th>
    Color
</th></div></div></div></div>
        "
`;
